libnd4j/include/ops/declarable/helpers/cpu/qr.cpp

```diff
@@ -0,0 +1,133 @@
+/*******************************************************************************
+  * Copyright (c) 2019-2020 Konduit K.K.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+//
+//  @author George A. Shulinok <sgazeos@gmail.com>
+//
+#include <ops/declarable/helpers/qr.h>
+#include <helpers/MmulHelper.h>
+#include <execution/Threads.h>
+#include <NDArrayFactory.h>
+
+namespace nd4j {
+namespace ops {
+namespace helpers {
+
+    template <typename T>
+    NDArray matrixMinor(NDArray& in, Nd4jLong col) {
+        NDArray m = in.ulike();
+
+        for (auto i = 0; i < col; i++)
+            m.t<T>(i,i) = T(1.f);
+        for (int i = col; i < m.rows(); i++)
+            for (int j = col; j < m.columns(); j++)
+                m.t<T>(i,j) = in.t<T>(i,j);
+        return m;
+    }
+
+/* m = I - v v^T */
+    template <typename T>
+    NDArray vmul(NDArray const& v, int n)
+    {
+        NDArray res('c', {n,n}, v.dataType()); // x = matrix_new(n, n);
+
+        auto interloop = PRAGMA_THREADS_FOR_2D {
+            for (int i = start_x; i < n; i += inc_x)
+                for (int j = start_y; j < n; j += inc_y)
+                    res.t<T>(i, j) = -2 * v.t<T>(i) * v.t<T>(j) + (i == j ? T(1) : T(0));
```

Comment:
Same. This function should not use .t method. It should operate on buffers instead for better performance.